philander/stc311_reg.py

```diff
@@ -0,0 +1,360 @@
+# TODO: all the comments
+
+"""Register definition for the STC311 battery gas gauge.
+    
+Definition of registers and default values for the
+above-mentioned chip.
+Externalized, just for clarity of the source code.
+"""
+__author__ = "Carl Bellgardt"
+__version__ = "0.1"
+__all__ = ["STC3115_Reg", "STC3117_Reg", "ChipType"]
+
+from enum import auto
+from .systypes import Enum
+
+
+class ChipType(Enum):
+    STC3115 = auto()
+    STC3117 = auto()
+
+
+class _ModeValues:
+    """This type describes different operating modes.
+    """
+    MODE_VMODE = 0x01  # 0: Mixed mode (Coulomb counter active); 1: Power saving voltage mode
+    MODE_ALM_ENA = 0x08  # Alarm function enable
+    MODE_GG_RUN = 0x10  # Standby / operating mode
+    MODE_FORCE_CC = 0x20  # Forces the relaxation timer to switch to the Coulomb counter (CC) state.
+    MODE_FORCE_VM = 0x40  # Forces the relaxation timer to switch to voltage mode (VM) state.
+
+
+class _CtrlValues:
+    """This type describes different control commands.
+    """
+    CTRL_IO0DATA = 0x01  # ALM pin status / ALM pin output drive
+    CTRL_GG_RST = 0x02  # resets the conversion counter GG_RST is a self-clearing bit
+    CTRL_GG_VM = 0x04  # Coulomb counter mode / Voltage mode active
+    CTRL_BATFAIL = 0x08  # Battery removal (BATD high).
+    CTRL_PORDET = 0x10  # Power on reset (POR) detection / Soft reset
+    CTRL_ALM_SOC = 0x20  # Set with a low-SOC condition
+    CTRL_ALM_VOLT = 0x40  # Set with a low-voltage condition
+    CTRL_DEFAULT = CTRL_IO0DATA
+
+
+class _STC311x_Reg:
+    def __init__(self, Param_dict):
```

Comment:
but the chip_type is need to know, which registers to use, and because the chip_type is only specified via the parameters (and those are definitively in open()), the registers can not be set correctly before open() is called. I'd say the registers need to depend on the parameters and thus the open method and cannot be set beforehand.

Reply:
Right, local variables (register shadows) cannot be **set** here. Still, they should be initialized, e.g. with zero, NULL or reasonable defaults. Remember that instance variables shall be "created/declared" and initialized in a class' constructor.